src/cargo/util/paths.rs

```diff
@@ -180,6 +180,42 @@ pub fn mtime(path: &Path) -> CargoResult<FileTime> {
     Ok(FileTime::from_last_modification_time(&meta))
 }
 
+/// Returns the maximum mtime of the given path, recursing into
+/// subdirectories, and following symlinks.
+pub fn mtime_recursive(path: &Path) -> CargoResult<FileTime> {
+    let meta = fs::metadata(path).chain_err(|| format!("failed to stat `{}`", path.display()))?;
+    if !meta.is_dir() {
+        return Ok(FileTime::from_last_modification_time(&meta));
+    }
+    let max_meta = walkdir::WalkDir::new(path)
+        .follow_links(true)
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter_map(|e| {
+            if e.path_is_symlink() {
+                // Use the mtime of both the symlink and its target, to
+                // handle the case where the symlink is modified to a
+                // different target.
+                let sym_meta = std::fs::symlink_metadata(e.path()).ok()?;
+                let sym_mtime = FileTime::from_last_modification_time(&sym_meta);
+                // Walkdir follows symlinks.
+                match e.metadata() {
+                    Ok(target_meta) => {
+                        let target_mtime = FileTime::from_last_modification_time(&target_meta);
+                        Some(sym_mtime.max(target_mtime))
```

Comment:
This seems a bit odd to me, if walkdir follows symlinks then how come we need to manually check the mtime of the symlink target? If the symlink points to a directory doesn't that mean we're doing the same thing as before where we're looking at the mtime of a directory which isn't too useful?

Reply:
If you have:
```
somedir/
    foo.txt
    bar.txt
    thing -> foo.txt
```

If you run a build, it will use the max mtime of those 3 files.  If you then modify `thing` to point to `bar.txt`, that will update the mtime of `thing`, and will trigger a rebuild.  If this didn't look at both the symlink and the target, it would only read the mtime of `foo.txt` and `bar.txt`, and Cargo would fail to trigger a rebuild.

If the `thing` symlink pointed to a directory, yes it will read the mtime of the directory.  But then, walkdir will traverse into that directory and scan all the contents.  Also, reading the mtime of the directory itself is important for deleting files.  Otherwise, Cargo wouldn't know that a file was deleted.

Reply:
Ah I see. I didn't actually realized until just before this that the mtime of directories was important. I was mostly wondering here why it dealt with both the target and the source since for file-based symlinks I think the symlink mtime itself should be appropriate, but even then I can see how it would matter if the file is located outside of the directory.

Reply:
> Also, reading the mtime of the directory itself is important for deleting files. Otherwise, Cargo wouldn't know that a file was deleted.

This seems like an error-prone way to find out a file was deleted. I think instead Cargo should cache the list of files in the directory (recursively) and consider the directory changed if any file that is in that cached list doesn't exist (or a new file exists that isn't on the list), in addition to the file timestamping.